docs/rfcs/027-pageserver-disaster-recovery.md

```diff
@@ -0,0 +1,198 @@
+
+### Design constraints
+
+The main design constraint is that the feature needs to be *simple* enough that
+the number of bugs are as low, and reliability as high as possible: the main
+goal of this endeavour is to achieve higher correctness than the pageserver.
+
+For the background process, we cannot afford a downtime of the timeline that is
+being cloned, as we don't want to restrict ourselves to offline tenants only.
+In the scenario where we want to recover from disasters or roll back to a
+historic lsn through support staff, downtimes are more affordable, and
+inevitable if the original had been subject to the corruption. Ideally, the
+two code paths would share code, so the solution would be designed for not
+requiring downtimes.
+
+### API endpoints
+
+The proposed design of the new pageserver API endpoint is:
+
+```
+/v1/tenant/:tenant_id/timeline/:timeline_id/recover?from_lsn=<from_lsn>&to_lsn=<lsn>&target=<target_timeline_id>
+```
+
+The `recover` endpoint answers to `POST` requests and issues a WAL recovery
+process. It takes the WAL of the specified timeline, and writes its results
+into a *different* timeline, the target timeline. This two timeline scheme
+ensures that we can stil serve requests and preserves immutability guarantees
+assumed by the [generation numbers RFC], as in corruption situations we might
+indeed end up with different content than the original timeline we are storing
+data from.
+
+Where the parameters are:
+
+* `:tenant_id`: the id of the tenant to recover
+* `:timeline_id`: the id of the timeline to recover
+* `from_lsn`: The lsn at the start of our recovery process.
+  Optional, if not specified the recovery process is fully WAL based.
+  If specified, the recovery process starts from data at the *main* timeline
+  at the specific lsn (maybe we still trust that lsn). In any case, the target
+  timeline must be less advanced than `from_lsn`.
+* `to_lsn`: the end of the lsn range we want to recover. As usual, it's an
+  exclusive limit. This allows playing back the WAL to a specific moment in the
+  past, if this is wanted/needed. E.g. there might be a data-corrupting
+  pageserver crash at lsn 1234 and we don't want that so we play back until
+  1233.
+* `target`: the timeline id we are writing into. The parameter is optional;
+  if no target timeline is provided, a new timeline is created.
+  One specifies either both `from_lsn` and `target`, or neither.
+  Writing into the same timeline is forbidden.
+
+The API call returns immediately after creation of the new timeline, and
+doesn't wait for the timeline's WAL replay to finish. The API call returns
+the id of the newly created timeline. The new timeline will appear in the
+usual listings and its lsn will increase until WAL replay has finished.
+As the caller knows the targeted lsn goal, this allows them to be informed
+about progress. There is no push notification for when the process has
+finished, any followup actions need to first poll for completion.
```

Comment:
> I don't fully understand this API. Does it work just like the normal branching, but replays the WAL from the beginning?

Yes.

> Is there a big difference to creating a timeline with ancestor_timeline_id = :timeline_id, ancestor_start_lsn = :to_lsn?

You wrote the difference above, which is that we replay the WAL instead of using the pageserver's internal datastructures to perform the branching. ideally, yes, the end result would be quite the same but the idea is that this recovery process is done when there has been a corruption.

We have discussed at the storage offsite whether the timeline should have the original timeline as its anchestor or not. Ideally we'd delete the original timeline however, otherwise it's staying around as a "ghost", eating up data. But then you'll need to be able to handle deleted timelines...

personally I think it would be best to not link to the original timeline as the anchestor but instead add the capability for the timeline to be the new "root" and instead handle things from there.

For recovering non-main branches

Reply:
> Personally I think it would be best to not link to the original timeline as the anchestor but instead add the capability for the timeline to be without anchestor, and handle it that way.

Note that we still need to remember timeline_id and start_lsn of the ancestor branch. This is needed to fetch WAL from safekeepers, which is stored under ancestor_timeline_id.